plugins/main/tap/tap.go

```diff
@@ -0,0 +1,464 @@
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/plugins/pkg/ip"
+	"github.com/containernetworking/plugins/pkg/ipam"
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+	"github.com/containernetworking/plugins/pkg/utils/sysctl"
+)
+
+type NetConf struct {
+	types.NetConf
+	MultiQueue bool   `json:"multiQueue"`
+	MTU        int    `json:"mtu"`
+	Mac        string `json:"mac,omitempty"`
+
+	RuntimeConfig struct {
+		Mac string `json:"mac,omitempty"`
+	} `json:"runtimeConfig,omitempty"`
+}
+
+// MacEnvArgs represents CNI_ARG
+type MacEnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(args *skel.CmdArgs) (*NetConf, string, error) {
+	n := &NetConf{}
+	if err := json.Unmarshal(args.StdinData, n); err != nil {
+		return nil, "", fmt.Errorf("failed to load netconf: %v", err)
+	}
+	if args.Args != "" {
+		e := MacEnvArgs{}
+		err := types.LoadArgs(args.Args, &e)
+		if err != nil {
+			return nil, "", err
+		}
+
+		if e.MAC != "" {
+			n.Mac = string(e.MAC)
+		}
+	}
+
+	if n.RuntimeConfig.Mac != "" {
+		n.Mac = n.RuntimeConfig.Mac
+	}
+
+	return n, n.CNIVersion, nil
+}
+
+func CloseOnExec(fd int) {
+	syscall.CloseOnExec(fd)
+}
+
+func setContainerSeBool() error {
+	output, err := exec.Command("cat", "/sys/fs/selinux/booleans/container_use_devices").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to run getsebool command %s: %v", string(output), err)
+	}
+	if strings.Contains(string(output), "off") {
+		output, err := exec.Command("setsebool", "-P", "container_use_devices", "true").CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("failed to run setsebool command %s: %v", string(output), err)
+		}
+	}
+	return nil
+}
+
+func createSelinuxTap(conf *NetConf, tmpName string) error {
+	if err := selinux.SetExecLabel("system_u:system_r:container_t:s0"); err != nil {
+		return fmt.Errorf("failed set socket label: %v", err)
+	}
+
+	minFDToCloseOnExec := 3
+	maxFDToCloseOnExec := 256
+	// we want to share the parent process std{in|out|err} - fds 0 through 2.
+	// Since the FDs are inherited on fork / exec, we close on exec all others.
+	for fd := minFDToCloseOnExec; fd < maxFDToCloseOnExec; fd++ {
+		CloseOnExec(fd)
+	}
+
+	tapDeviceArgs := []string{"tuntap", "add", "mode", "tap", "name", tmpName}
+	if conf.MultiQueue {
+		tapDeviceArgs = append(tapDeviceArgs, "multi_queue")
+	}
+
+	output, err := exec.Command("ip", tapDeviceArgs...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to run command %s: %v", output, err)
+	}
```

Comment:
I'm unsure how serious this PR is; we shouldn't - IMO - use the ip tool, but either use https://github.com/songgao/water, or use syscalls to create the tap. 

(I don't remember if netlink allowed for tap creation or if I'm confusing it w/ macvtap).

Reply:
it's possible to create with netlink you can see that we do it when we are not running on selinux system.

and we use the ip command because we need to start a second binary so the setExecLabel will be applied to them.

another solution will be to create a different binary with netlink but that will make the testing (you need to have the binary compiled) and the ship (another binary in the CNI repo with is just a helper) more complicated.

but I think we are open to suggestions

Reply:
I see.

I'd still vote for not using the ip tool despite its higher cost of having to ship the tool . 

For the sake of the discussion this is fine.

Reply:
We need to use a new process to make use of the new label.
One option is to use some external one like 'ip' (which we assume will be on the nodes). Another option would be for the plugin to call itself again (with the new label) and use the golang libs for this (I assume netlink would do in this case).

Reply:
> We need to use a new process to make use of the new label.

I'm aware you need a new process. I'm just saying you don't need IP for that. I think using a smaller binary with a netlink implementation that creates a tap device setting the owner / group / mtu would be enough.

Reply:
Changed this to use a binary with netlink.
To avoid having multiple binaries the plugin binary will be used for that.

Reply:
I like where this landed, thanks for the discussion, I really like the bin calling itself with the proper selinux, seems really all-in-one.

Reply:
> we are not running on selinux system

CNI plugins might run on systems that don't use selinux.

Reply:
This is fine. The check for selinux genenforce function will just return false, and we'll proceed with the default.